vpr/src/route/rr_graph_storage.h

```diff
@@ -0,0 +1,578 @@
+        VTR_ASSERT_SAFE(ret < last_edge(id));
+        return ret;
+    }
+
+    // Get the destination node for the specified edge.
+    RRNodeId edge_sink_node(const RREdgeId& edge) const {
+        return edge_dest_node_[edge];
+    }
+
+    // Get the destination node for the iedge'th edge from specified RRNodeId.
+    //
+    // This method should generally not be used, and instead first_edge and
+    // last_edge should be used.
+    RRNodeId edge_sink_node(const RRNodeId& id, t_edge_size iedge) const {
+        return edge_sink_node(edge_id(id, iedge));
+    }
+
+    // Get the switch used for the specified edge.
+    short edge_switch(const RREdgeId& edge) const {
+        return edge_switch_[edge];
+    }
+
+    // Get the switch used for the iedge'th edge from specified RRNodeId.
+    //
+    // This method should generally not be used, and instead first_edge and
+    // last_edge should be used.
+    short edge_switch(const RRNodeId& id, t_edge_size iedge) const {
+        return edge_switch(edge_id(id, iedge));
+    }
+
+    /******************
+     * Fan-in methods *
+     ******************/
+
+    /* Init per node fan-in data.  Should only be called after all edges have
+     * been allocated */
+    void init_fan_in();
+
+    /* Retrieve fan_in for RRNodeId, init_fan_in must have been called first. */
+    t_edge_size fan_in(RRNodeId id) {
+        return fan_in_[id];
+    }
+
+  private:
+    friend struct edge_swapper;
+    friend class edge_sort_iterator;
+    friend class edge_compare_dest_node;
+    friend class edge_compare_src_node_and_configurable_first;
+
+    // Take allocated edges in edge_src_node_/ edge_dest_node_ / edge_switch_
+    // sort, and assign the first edge for each
+    void assign_first_edges();
+
+    // Verify that first_edge_ array correctly partitions rr edge data.
+    bool verify_first_edges() const;
+
+    /*****************
+     * Graph storage
+     *
+     * RR graph storage generally speaking includes two types of data:
+     *  - **Hot** data is used in the core place and route algorithm.
+     *    The layout and size of data stored in **hot** storage will likely
+     *    have a significant affect on router performance.  Any changes to
+     *    hot data should carefully considered and measured, see
+     *    README.developers.md.
+     *
+     *  - **Cold** data is not used in the core place and route algorithms,
+     *    so it isn't as critical to be measure the affects, but it **is**
+     *    important to watch the overall memory usage of this data.
+     *
+     *****************/
+
+    // storage_ stores the core RR node data used by the router and is **very**
+    // hot.
+    vtr::vector<RRNodeId, t_rr_node_data, vtr::aligned_allocator<t_rr_node_data>> storage_;
+
+    // The PTC data is cold data, and is generally not used during the inner
+    // loop of either the placer or router.
+    vtr::vector<RRNodeId, t_rr_node_ptc_data> ptc_;
```

Comment:
prefix these with node_ for consistency.

Reply:
done